CCIP-4948: Adding new test for disable and enable lane

b-gopalswami opened the pull request:
Adding test to verify the op changeset workflow to disable and re-enable the lane.

Tests it will cover: TC001 & TC002
https://smartcontract-it.atlassian.net/wiki/spaces/QA/pages/1262846210/CCIP+V1.6+Changeset+Test+Plan

b-gopalswami committed:
CCIP-4948: Adding new test for disable and enable lane

b-gopalswami commented on deployment/ccip/changeset/cs_chain_contracts.go:
```diff
@@ -705,12 +705,10 @@ func UpdateOffRampSourcesChangeset(e deployment.Environment, cfg UpdateOffRampSo
 		var args []offramp.OffRampSourceChainConfigArgs
 		for source, update := range updates {
 			router := common.HexToAddress("0x0")
-			if update.IsEnabled {
```
router can't be 0x0 address as Offramp expects valid router address ow it will throw ZeroAddressNotAllowed error message. https://github.com/smartcontractkit/chainlink/blob/ba148ac9e92b2dfa24693d84467d4c55958501f5/contracts/src/v0.8/ccip/offRamp/OffRamp.sol#L994

AnieeG replied:
If it's not possible to disable a router in offRamp, the parameter IsEnabled should be removed

b-gopalswami replied:
I think it is still needed to make here https://github.com/smartcontractkit/chainlink/blob/48c7cb700a1e2fae55e23182561a9c3445c81b0d/deployment/ccip/changeset/cs_chain_contracts.go#L718

b-gopalswami committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into ccip-4948

b-gopalswami committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into ccip-4948

b-gopalswami committed:
Send request before disabling the lane

b-gopalswami committed:
lint fix

fix newline

b-gopalswami committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into ccip-4948

b-gopalswami committed:
revert gethwrappers change

AnieeG commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -482,6 +482,45 @@ func AddLane(
 	require.NoError(t, err)
 }
 
+func UpdateLane(
```
Comment please

AnieeG commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -482,6 +482,45 @@ func AddLane(
+) {
+	var err error
+	var apps []commoncs.ChangesetApplication
+	apps = append(apps, commoncs.ChangesetApplication{
```
To disable a lane, I think we need to disable the router to ramp connection, otherwise router will still route the requests to ramps

b-gopalswami committed:
review comments

b-gopalswami committed:
Disabling fee quoter as well

review comments

rename

Add test to CI and add enabledConfig flag

remove unwanted sql

add variables locally

AnieeG commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -482,6 +482,87 @@ func AddLane(
+							dest: isEnabled,
+						},
+					},
+					// offramp update on dest chain
+					src: {
+						OffRampUpdates: map[uint64]bool{
+							dest: isEnabled,
+						},
```
your comment does not match the selector key , your updating `src`.
Also this part is not required. We don't want to disable offRamp on dest router as the inflight requests and all manual exec requests will be stuck if offRamp is removed from dest router.

AnieeG commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -482,6 +482,87 @@ func AddLane(
+			},
+		},
+		{
+			Changeset: commoncs.WrapChangeSet(changeset.UpdateFeeQuoterDestsChangeset),
+			Config: changeset.UpdateFeeQuoterDestsConfig{
+				UpdatesByChain: map[uint64]map[uint64]fee_quoter.FeeQuoterDestChainConfig{
+					src: {
+						dest: defaultFeeQuoterDestConfig,
+					},
+				},
+			},
+		},
```
Is this disabling the dest config from feequoter? no right?

AnieeG commented on deployment/ccip/changeset/cs_chain_contracts.go:
```diff
@@ -1124,7 +1124,7 @@ func DefaultFeeQuoterDestChainConfig() fee_quoter.FeeQuoterDestChainConfig {
 	evmFamilySelector, _ := hex.DecodeString("2812d52c")
-	return fee_quoter.FeeQuoterDestChainConfig{
+	feeQuoterDestChainConfig := fee_quoter.FeeQuoterDestChainConfig{
 		IsEnabled:                         true,
```
this should be changed based on your input

AnieeG replied:
```suggestion
		IsEnabled:                         configEnabled,
```

AnieeG commented on deployment/ccip/changeset/cs_chain_contracts.go:
```diff
@@ -1143,4 +1143,8 @@ func DefaultFeeQuoterDestChainConfig() fee_quoter.FeeQuoterDestChainConfig {
 		NetworkFeeUSDCents:                1,
 		ChainFamilySelector:               [4]byte(evmFamilySelector),
 	}
+	if !configEnabled {
+		feeQuoterDestChainConfig.IsEnabled = false
+	}
+	return feeQuoterDestChainConfig
```
Change this directly no need for condition check
```suggestion
	
	return feeQuoterDestChainConfig
```

AnieeG commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -482,6 +482,65 @@ func AddLane(
+			},
+		},
+		{
+			Changeset: commoncs.WrapChangeSet(changeset.UpdateOffRampSourcesChangeset),
+			Config: changeset.UpdateOffRampSourcesConfig{
+				UpdatesByChain: map[uint64]map[uint64]changeset.OffRampSourceUpdate{
+					dest: {
+						src: {
+							IsEnabled:  false,
+							TestRouter: isTestRouter,
+						},
+					},
+				},
+			},
+		},
+	}
```
@makramkd would it affect the inflight requests?

makramkd replied:
Yeah, no offramp changes should be made

b-gopalswami replied:
In my observation, if no offramp changes are made, then in-flight requests are getting delivered. Let's say, we made the offramp changes as well, in that case the in-flight requests are not getting delivered. Haven't seen reverting as well not sure If I need to wait longer for that.

So, we should not make offramp changes when there is in-flight request. Will remove this part.

makramkd replied:
>  if no offramp changes are made, then in-flight requests are getting delivered.

Yeah, just to be clear this is the desired behavior.

AnieeG commented on integration-tests/smoke/ccip/ccip_disable_lane_test.go:
```diff
@@ -0,0 +1,137 @@
+	assertSendRequestReverted(chainB, chainA, e.Chains[chainB].Users[0])
+
+	// send message in other lanes and ensure they are delivered
+	go func() {
+		assertRequestSent(chainA, chainC, e.Chains[chainA].Users[1])
+		assertRequestSent(chainC, chainA, e.Chains[chainC].Users[1])
+		assertRequestSent(chainB, chainC, e.Chains[chainB].Users[1])
+		assertRequestSent(chainC, chainB, e.Chains[chainC].Users[1])
+	}()
+	// disable lanes between A & C and C & B while requests are getting sent
+	testhelpers.RemoveLane(t, &tenv, chainA, chainC, false)
+	testhelpers.RemoveLane(t, &tenv, chainC, chainA, false)
+	testhelpers.RemoveLane(t, &tenv, chainB, chainC, false)
+	testhelpers.RemoveLane(t, &tenv, chainC, chainB, false)
```
This could lead to flakiness I think. How about -
1. Just use one lane and send 10 requests
2. Remove that lane.
3. Ensure all requests are delivered even after removing lane

b-gopalswami committed:
fix test

b-gopalswami committed:
fix lint

AnieeG commented on integration-tests/smoke/ccip/ccip_disable_lane_test.go:
```diff
@@ -0,0 +1,155 @@
+	testhelpers.ConfirmExecWithSeqNrsForAll(t, e, state, expectedSeqNumExec, startBlocks)
+
+	// disable lane B -> A
+	pairs = append(pairs, testhelpers.SourceDestPair{
+		SourceChainSelector: chainB,
+		DestChainSelector:   chainA,
+	})
+	testhelpers.RemoveLane(t, &tenv, chainB, chainA, false)
+	assertSendRequestReverted(chainB, chainA, e.Chains[chainB].Users[0])
```
can be removed, repetitive scenario

AnieeG commented on integration-tests/smoke/ccip/ccip_disable_lane_test.go:
```diff
@@ -0,0 +1,155 @@
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for range 5 {
```
As var in the beginning of this test

AnieeG commented on integration-tests/smoke/ccip/ccip_disable_lane_test.go:
```diff
@@ -0,0 +1,155 @@
+		SourceChainSelector: chainA,
+		DestChainSelector:   chainC,
+	})
+	testhelpers.RemoveLane(t, &tenv, chainA, chainC, false)
```
How are you ensuring that Remove Lane will not finish execution while the previous requests are yet to be sent?

b-gopalswami replied:
Right chances of sending request after removal is there. I think the we can fire the request first and remove the lane.

b-gopalswami committed:
review comments

remove if
